File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/NotificationsResource.java
package work.socialhub.kmastodon.api;

import mastodon4j.Range;
import mastodon4j.entity.Alert;
import mastodon4j.entity.Notification;
import mastodon4j.entity.Subscription;
import mastodon4j.entity.share.Response;

import java.util.List;

/**
 * @author hecateball
 */
public interface NotificationsResource {

    /**
     * Fetching a user's notifications.
     *
     * @return a list of Notifications for the authenticated user
     */
    public Response<Notification[]> getNotifications(
            Range range,
            List<String> types,
            List<String> excludeTypes,
            Long id);

    /**
     * Getting a single notification.
     *
     * @param id
     * @return the Notification
     */
    public Response<Notification> getNotification(String id);

    /**
     * Clearing notifications. Deletes all notifications from the Mastodon server for the authenticated user.
     */
    public Response<Void> clearNotifications();


    /**
     * Get current subscription.
     *
     * @return subscription info.
     */
    public Response<Subscription> getSubscription();

    /**
     * Subscribe to push notifications.
     * Register service worker endpoint.
     *
     * @return subscription info that registered.
     */
    public Response<Subscription> pushSubscription(
            String endpoint,
            String p256dh,
            String auth,
            Alert alert);

    /**
     * Change types of notifications
     *
     * @return subscription info.
     */
    public Response<Subscription> editSubscription(
            Alert alert);
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/api/NotificationsResource.kt
package work.socialhub.kmastodon.api

import work.socialhub.kmastodon.api.request.notifications.NotificationsEditSubscriptionRequest
import work.socialhub.kmastodon.api.request.notifications.NotificationsNotificationRequest
import work.socialhub.kmastodon.api.request.notifications.NotificationsNotificationsRequest
import work.socialhub.kmastodon.api.request.notifications.NotificationsPostSubscriptionRequest
import work.socialhub.kmastodon.api.response.Response
import work.socialhub.kmastodon.api.response.ResponseUnit
import work.socialhub.kmastodon.api.response.notifications.NotificationsEditSubscriptionResponse
import work.socialhub.kmastodon.api.response.notifications.NotificationsNotificationResponse
import work.socialhub.kmastodon.api.response.notifications.NotificationsNotificationsResponse
import work.socialhub.kmastodon.api.response.notifications.NotificationsPostSubscriptionResponse
import work.socialhub.kmastodon.api.response.notifications.NotificationsSubscriptionResponse
import kotlin.js.JsExport

@JsExport
interface NotificationsResource {

    /**
     * Fetching a user's notifications.
     */
    suspend fun notifications(
        request: NotificationsNotificationsRequest
    ): Response<Array<NotificationsNotificationsResponse>>

    @JsExport.Ignore
    fun notificationsBlocking(
        request: NotificationsNotificationsRequest
    ): Response<Array<NotificationsNotificationsResponse>>

    /**
     * Getting a single notification.
     */
    suspend fun notification(
        request: NotificationsNotificationRequest
    ): Response<NotificationsNotificationResponse>

    @JsExport.Ignore
    fun notificationBlocking(
        request: NotificationsNotificationRequest
    ): Response<NotificationsNotificationResponse>

    /**
     * Clearing notifications. Deletes all notifications
     * from the Mastodon server for the authenticated user.
     */
    suspend fun clearNotifications(
    ): ResponseUnit

    @JsExport.Ignore
    fun clearNotificationsBlocking(
    ): ResponseUnit

    /**
     * Get current subscription.
     */
    suspend fun subscription(
    ): Response<NotificationsSubscriptionResponse>

    @JsExport.Ignore
    fun subscriptionBlocking(
    ): Response<NotificationsSubscriptionResponse>

    /**
     * Subscribe to push notifications.
     * Register service worker endpoint.
     */
    suspend fun pushSubscription(
        request: NotificationsPostSubscriptionRequest
    ): Response<NotificationsPostSubscriptionResponse>

    @JsExport.Ignore
    fun pushSubscriptionBlocking(
        request: NotificationsPostSubscriptionRequest
    ): Response<NotificationsPostSubscriptionResponse>

    /**
     * Change types of notifications
     */
    suspend fun editSubscription(
        request: NotificationsEditSubscriptionRequest
    ): Response<NotificationsEditSubscriptionResponse>

    @JsExport.Ignore
    fun editSubscriptionBlocking(
        request: NotificationsEditSubscriptionRequest
    ): Response<NotificationsEditSubscriptionResponse>
}
